src/jupyterlabcontroller/services/lab.py

```diff
@@ -69,175 +206,339 @@ async def create_lab(
+            if lab.state.status == LabStatus.TERMINATED:
+                lab.state = None
+        else:
+            raise OperationConflictError(username)
+        if lab.state and lab.state.status != LabStatus.TERMINATED:
+            msg = f"Deleting lab for {username} failed"
+            raise LabDeletionError(msg, username)
 
-        # Retrieve the secrets that will be used to construct the lab secret.
-        # Generate a set of names first so that we retrieve each secret only
-        # once.
-        self._logger.info("Retrieving secret data")
-        pull_secret = None
-        try:
-            secret_data = await self._gather_secret_data(user)
-            if self._config.pull_secret:
-                pull_secret = await self._storage.read_secret(
-                    self._config.pull_secret, self._metadata.namespace
-                )
-        except MissingSecretError as e:
-            e.user = username
-            raise
+    def events_for_user(self, username: str) -> AsyncIterator[ServerSentEvent]:
+        """Iterator over the events for a user.
 
-        # Build the objects that make up the user's lab.
-        objects = self._builder.build_lab(
-            user=user,
-            lab=lab,
-            image=image,
-            secrets=secret_data,
-            pull_secret=pull_secret,
-        )
+        Parameters
+        ----------
+        username
+            Username for which to retrieve events.
 
-        # Create the lab objects in Kubernetes.
-        self._logger.info("Creating new lab")
-        await self._storage.create(objects)
-        await self._lab_state.publish_pod_creation(
-            username, "Created Kubernetes objects for user lab", 30
-        )
+        Yields
+        ------
+        Event
+            Events for that user until a completion or failure event is seen.
 
-        # Return the URL where the lab will be listening after it starts.
-        return self._builder.build_internal_url(user.username, lab.env)
+        Raises
+        ------
+        UnknownUserError
+            Raised if there is no event stream for this user.
+        """
+        if username not in self._labs:
+            raise UnknownUserError(f"Unknown user {username}")
 
-    async def delete_lab(self, username: str) -> None:
-        """Delete the lab environment for the given user.
+        async def iterator() -> AsyncIterator[ServerSentEvent]:
+            async for event in self._labs[username].events:
+                yield event.to_sse()
+
+        return iterator()
+
+    async def get_lab_state(self, username: str) -> UserLabState | None:
+        """Get lab state for a user.
+
+        This method underlies the API called by JupyterHub to track whether
+        the user's lab still exists. We want to update that state after
+        Kubernetes changes faster than our reconciliation cycle, so ask
+        Kubernetes directly for the pod phase each time we're asked for the
```

Comment:
How often is this, potentially?  Do we ever need to worry about it?

Reply:
Every Kubernetes upgrade deletes a bunch of pods, so if we don't do this, all those affected users have to wait for the hour-long reconciliation period before they can spawn again. JupyterHub calls this when the user goes to the spawner page, so this means that the user can recover effectively instantly.